### PR TITLE
Implement the xwayland shell protocol

### PR DESCRIPTION
**Pull request by colinmarc**
Depends on #1350 (but only trivially).

The xwayland shell protocol moves the responsibility of setting the "xwayland_surface" role on the wl_surface from the x11 side of the connection to the wayland side. Among other things, this avoids the xwm racing the lifecycle of the surface.

Since we can check the serial with a commit hook, this also removes the need for a ugly idle callback in the `Xwm`, and may allow for better separation of concerns in the future, for example moving the association state tracking out of Xwm.

Without the idle callback, we may be able to remove the hard dependency on calloop in the xwayland stack (which is my ulterior motive). It's still required for `send_selection` - I'll attempt to figure those issues out in another PR.

### CODE REVIEW

File: src/xwayland/xwm/mod.rs
Comment (colinmarc) [inline src/xwayland/xwm/mod.rs:2129]: This is really unfortunate. We should only be setting this in the `else` below, but that would mean double-borrowing `state`. I stared at this for tens of minutes without seeing a way around it.

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [inline src/xwayland/xwm/mod.rs:2153]: I think if you changed this to `cloned`, you can avoid the double borrow.

(Referencing https://github.com/Smithay/smithay/pull/1351/commits/64ebcfa9a3229c6d916f194481a2389ac0c9fcec#r1510053974)

File: anvil/src/shell/x11.rs
Comment (Drakulix) [inline anvil/src/shell/x11.rs:54]: I don't think we want a second trait and force downstream to awkwardly forward calls like this. I think it is fine to just drop the old matching logic and require a recent enough Xwayland version supporting this protocol.

File: src/xwayland/xwm/mod.rs
Comment (colinmarc) [reply src/xwayland/xwm/mod.rs:2153]: I can check later (mobile), but I'm pretty sure the issue is using `xwm` again after using `&mut state` at all, even in a temporary variable - there's an implicit `drop(xwm)` which makes the code work right now.

File: anvil/src/shell/x11.rs
Comment (colinmarc) [reply anvil/src/shell/x11.rs:54]: This hack predates my PR, I just made it explicit. I'll try to explain. It has nothing to do with the old matching logic.

This is necessary because Anvil has an `EventLoop<CalloopData>` but a different state object (`AnvilState`), on which most handlers are implemented. Because `Xwm` requires `LoopHandle<D> where D: XwmHandler`, the existing code implemented `XwmHandler` directly on `CalloopData` instead of `AnvilState`. But then it had to call into the Xwm from one of those other handlers on `AnvilState`:

https://github.com/Smithay/smithay/blob/418190e4992ce642e6bac873307d4fc4fa9a1e89/anvil/src/shell/mod.rs#L142

I changed the commit hook to take a `D` as a parameter, so that broke. I actually think this hack is cleaner, in a way, although it's terrible on its face.

Note that this all goes away if remove the tight integration with calloop that xwm has. I would propose doing that by adding a `poll_fd()` and a `dispatch` method, mirroring the wayland side.

By the second trait, do you mean `XWaylandShellHandler`? I think having that separated makes sense, since it's just for the protocol itself. And that means people can supply their own xwm implementation.

Comment (Drakulix) [reply anvil/src/shell/x11.rs:54]: > Note that this all goes away if remove the tight integration with calloop that xwm has. I would propose doing that by adding a poll_fd() and a dispatch method, mirroring the wayland side.

Good idea.

> By the second trait, do you mean XWaylandShellHandler? I think having that separated makes sense, since it's just for the protocol itself. And that means people can supply their own xwm implementation.

Then maybe we need to split it. But the duplicated x-wm callbacks are not nice to use.

Comment (colinmarc) [reply anvil/src/shell/x11.rs:54]: Sorry, I'm a bit lost. What do you mean by duplicated callbacks?

Comment (Drakulix) [reply anvil/src/shell/x11.rs:54]: Having the `XWaylandShellHandler` call into `XwmHandler`.

Comment (colinmarc) [reply anvil/src/shell/x11.rs:54]: > Having the XWaylandShellHandler call into XwmHandler.

I don't see where it does that. The two traits are independent, except that `XwmHandler` requires `D` to also implement `XWaylandShellHandler`, since `X11wm` uses both to do the matching. (I should actually move the constraint to `X11Wm` to make that clearer, on reflection, but that's a trivial change).

What's happening here (where this comment is) is that we're implementing `XWaylandShellHandler` for both `CalloopData<BackendData>` and `AnvilState<BackendData`. The forwarding is just between two implementations of the same trait, because of an interaction between `X11Wm` weirdness and `Anvil` weirdness. There's no cross-pollination between `XWaylandShellHandler` and `XwmHandler`.

I'm really sorry if I'm just being thick! I have a head cold at the moment and having trouble focusing :)

File: anvil/src/shell/mod.rs
Comment (colinmarc) [inline anvil/src/shell/mod.rs:141]: Referencing the discussion about trait forwarding below: https://github.com/Smithay/smithay/pull/1351/files/79648e06a287286f00520afc3d12efbdd4c38277..64ebcfa9a3229c6d916f194481a2389ac0c9fcec#r1511009746

Removing this hack (because of the addition of the `D` parameter to `X11Wm::commit_hook`) is what made the trait forwarding necessary, just for anvil.

File: anvil/src/shell/x11.rs
Comment (Drakulix) [reply anvil/src/shell/x11.rs:54]: > I don't see where it does that.

Ah my god, the diff view had me confused... sorry!

Yeah it was the two `XwmHandler` implementations (once for `CalloopData<AnvilState>` and the other for `AnvilState`), that were throwing me off.

I don't think we can really keep those in anvil, even if a normal compositor wouldn't have to do this, because this encourages people to just copy this hacky approach.

We should get rid of CalloopData in anvil, is that something you feel like possibly tackling?

Comment (colinmarc) [reply anvil/src/shell/x11.rs:54]: > I should actually move the constraint to X11Wm to make that clearer, on reflection, but that's a trivial change

Addressed in 6c675b5.

> We should get rid of CalloopData in anvil, is that something you feel like possibly tackling?

Between that and removing `X11Wm`'s dependency on `LoopHandle<D>`, and given that either would free us of this hack - I would rather work on the latter. :sweat_smile:

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [inline src/xwayland/xwm/mod.rs:1282]: If we require this here, we don't really support compositors using the old association method, do we?

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [inline src/xwayland/xwm/mod.rs:2138]: Referencing the previous comment.

File: src/xwayland/xwm/mod.rs
Comment (colinmarc) [reply src/xwayland/xwm/mod.rs:1282]: We require compositors to support the new association method. Whether they also handle the legacy method for old xwayland versions by checking for id-based associations in an idle callback (as I mention in the comment) is optional.

File: anvil/src/shell/x11.rs
Comment (Drakulix) [reply anvil/src/shell/x11.rs:54]: So what needs to be done to get rid of this double implementation in anvil?

Comment (colinmarc) [reply anvil/src/shell/x11.rs:54]: I took a look at changing `Xwm`'s hard dependency on calloop, and it was a bit too tangled up for me. I would love some help with that, or with ditching `CalloopData` in anvil. Either would remove the need for this hack.

**Comment by colinmarc** [comment]
Rebased and removed the anvil hack. This is only blocked on #1350 now (although I could separate it out if you really want).

File: anvil/src/state.rs
Comment (Drakulix) [inline anvil/src/state.rs:623]: I guess we could move this into start_xwayland, just like the XwaylandKeyboardGrab protocol?

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [inline src/xwayland/xwm/mod.rs:23]: I don't think this description makes it clear to the user, what they are supposed to do. You describe an internal mechanism from the users perspective. Open questions are:
- Are they supposed to match the surfaces themselves? Or does smithay match them?
- Does this module work without the XwaylandShell protocol?

**Review by Drakulix — COMMENTED** [review]
Some nitpicks, overall this is quite clean and I like the implementation a lot.

File: anvil/src/state.rs
Comment (colinmarc) [reply anvil/src/state.rs:625]: It means making that field an `Option`. What if I just move `XwaylandKeyboardGrab` protocol out here?

File: src/xwayland/xwm/mod.rs
Comment (colinmarc) [reply src/xwayland/xwm/mod.rs]: I rewrote this part, let me know what you think.

File: anvil/src/state.rs
Comment (Drakulix) [reply anvil/src/state.rs:625]: sure

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [reply src/xwayland/xwm/mod.rs:2179]: I guess you could re-borrow the `xwm` after getting the surface, but meh.. this is fine.

File: src/xwayland/xwm/surface.rs
Comment (Drakulix) [inline src/xwayland/xwm/surface.rs:251]: I am not sure this is right for override-redirect windows. We don't set the mapped_onto for those, but those can still be mapped/unmapped in which case Xwayland will remove/re-create the surface.

In general it is not a correct assumption, that an X-window will only have a single wl_surface associated during it's whole lifetime.

Comment (colinmarc) [reply src/xwayland/xwm/surface.rs:251]: I'm not sure about what the context for this function is, but it seems like it was wrong before as well?

Comment (Drakulix) [reply src/xwayland/xwm/surface.rs:251]: yeah I guess just checking for the surface to exist would be enough?

File: src/xwayland/xwm/mod.rs
Comment (Drakulix) [reply src/xwayland/xwm/mod.rs]: sounds good now!

File: src/xwayland/xwm/surface.rs
Comment (colinmarc) [reply src/xwayland/xwm/surface.rs]: Changed, but I'm not sure I understand. :smile:

File: anvil/src/state.rs
Comment (colinmarc) [reply anvil/src/state.rs:625]: Done.

**Comment by colinmarc** [comment]
CI failures appear unrelated.

File: src/xwayland/xwm/surface.rs
Comment (Drakulix) [reply src/xwayland/xwm/surface.rs]: You did in fact understand me correctly. :D

**Comment by Drakulix** [comment]
CI failure should be fixed by https://github.com/Smithay/smithay/pull/1424/files, can you rebase?

**Review by Drakulix — APPROVED** [review]
Good to go, once CI passes. Thanks for putting in all the work!!

**Pull request merged**
